File: swift/Tests/WarehouseTests/MoneyTests.swift
import Testing
import Warehouse

@Suite("Money")
struct MoneyTests {
  @Test func rejectsNegativeAmount() {
    expectInvalidOrder(
      { _ = try Money(minorUnits: -1, currency: "USD") },
      contains: "non-negative"
    )
  }

  @Test func rejectsMalformedCurrency() {
    expectInvalidOrder(
      { _ = try Money(minorUnits: 1, currency: "usd") },
      contains: "currency"
    )
  }

  @Test func addSumsSameCurrency() throws {
    let sum = try usd(150).add(try usd(275))
    #expect(sum == try usd(425))
  }

  @Test func addRejectsCurrencyMismatch() {
    expectInvalidOrder(
      {
        _ = try Money(minorUnits: 100, currency: "USD").add(
          try Money(minorUnits: 100, currency: "EUR")
        )
      },
      contains: "mismatch"
    )
  }

  @Test func timesScalesAmount() throws {
    #expect(try usd(250).times(3) == try usd(750))
  }

  @Test func timesRejectsNegativeMultiplier() {
    expectInvalidOrder(
      { _ = try usd(1).times(-2) },
      contains: "multiplier"
    )
  }

  @Test func equalityIsValueBased() throws {
    #expect(
      try Money(minorUnits: 10, currency: "EUR")
        == Money(minorUnits: 10, currency: "EUR")
    )
  }

  @Test func acceptsIsoStyleZzz() throws {
    #expect(try Money(minorUnits: 0, currency: "ZZZ").currency == "ZZZ")
  }

  @Test func rejectsAboveSharedMaximum() {
    expectInvalidOrder(
      { _ = try Money(minorUnits: moneyMinorUnitsMax + 1, currency: "USD") },
      contains: "exceeds"
    )
  }

  @Test func addRejectsOverflow() {
    expectInvalidOrder(
      {
        let maxMoney = try Money(minorUnits: moneyMinorUnitsMax, currency: "USD")
        _ = try maxMoney.add(try Money(minorUnits: 1, currency: "USD"))
      },
      contains: "overflows"
    )
  }

  @Test func timesZeroIsZero() throws {
    #expect(try usd(250).times(0) == try usd(0))
  }
}

File: csharp/tests/Warehouse.UnitTests/MoneyTests.cs
using Warehouse.Domain;

namespace Warehouse.UnitTests;

public sealed class MoneyTests
{
    [Fact]
    public void RejectsNegativeAmount()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => new Money(-1, "USD"));
        Assert.Contains("non-negative", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsMalformedCurrency()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => new Money(1, "usd"));
        Assert.Contains("currency", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void AddSumsSameCurrency()
    {
        Assert.Equal(new Money(425, "USD"), new Money(150, "USD").Add(new Money(275, "USD")));
    }

    [Fact]
    public void AddRejectsCurrencyMismatch()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(
            () => new Money(100, "USD").Add(new Money(100, "EUR")));
        Assert.Contains("mismatch", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void TimesScalesAmount()
    {
        Assert.Equal(new Money(750, "USD"), new Money(250, "USD").Times(3));
    }

    [Fact]
    public void TimesRejectsNegativeMultiplier()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => new Money(1, "USD").Times(-2));
        Assert.Contains("multiplier", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void EqualityIsValueBased()
    {
        Assert.Equal(new Money(10, "EUR"), new Money(10, "EUR"));
    }

    [Fact]
    public void AcceptsIsoStyleZzz()
    {
        Assert.Equal("ZZZ", new Money(0, "ZZZ").Currency);
    }

    [Fact]
    public void RejectsAboveSharedMaximum()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(
            () => new Money(9_007_199_254_740_992L, "USD"));
        Assert.Contains("exceeds", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void AddRejectsOverflow()
    {
        var max = new Money(Money.MaxMinorUnits, "USD");
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => max.Add(new Money(1, "USD")));
        Assert.Contains("overflows", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void TimesRejectsOverflow()
    {
        var amount = new Money(Money.MaxMinorUnits, "USD");
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => amount.Times(2));
        Assert.Contains("overflows", error.Message, StringComparison.Ordinal);
    }
}
